packages/workspace/src/browser/workspace-service.ts

```diff
@@ -21,21 +21,30 @@ import { FileSystemWatcher } from '@theia/filesystem/lib/browser';
 import { WorkspaceServer } from '../common';
 import { WindowService } from '@theia/core/lib/browser/window/window-service';
 import { FrontendApplication, FrontendApplicationContribution } from '@theia/core/lib/browser';
+import { Disposable, Emitter, Event } from '@theia/core/lib/common';
 import { Deferred } from '@theia/core/lib/common/promise-util';
 import { ILogger } from '@theia/core/lib/common/logger';
 import { WorkspacePreferences } from './workspace-preferences';
 
+export interface WorkspaceDescription {
```

Comment:
This structure is not very pretty and I don't think is needed, too.
Whether a workspace file pointed to a folder or config file is only important when computing the roots.
It can be determined when needed and doesn't need to be stored.

Reply:
You are right, but I am completely following your suggestion here. the `workspaceConfig` has been removed from this structure, however `workspaceFolder` (used to be called `root`), and `roots` stay. The reason, like i mentioned in another thread, is that users can possibly have more than one roots in the workspace and remove the workspace folder after.

Reply:
Not sure I understand. Let me try to outline how I understand this:
We have a workspace description that is a URI pointing to a folder or a file. 
If it is a folder we have a single root (that folder)
If it's a file we read its contents containing a list of folders, which will be the roots.

So all we need to expose as API is 'roots', because external services should not need to know how this service is storing or obtaining the information. And internally we can initially once read the workspace description to get the roots and then watch the file for changes.